### ISSUE TITLE
Test failures under PHP 7.3

### ISSUE THREAD

**Issue by Majkl578** [Roave/BetterReflection#440]
The following errors & failures are reported under PHP 7.3:
```
There were 10 errors:

1) Roave\BetterReflectionTest\Reflection\ReflectionParameterTest::testGetDocBlockTypeStrings
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindParameterType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionParameter.php:347
/tmp/BetterReflection/src/Reflection/ReflectionParameter.php:330
/tmp/BetterReflection/test/unit/Reflection/ReflectionParameterTest.php:194

2) Roave\BetterReflectionTest\Reflection\ReflectionParameterTest::testGetDocBlockTypes
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindParameterType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionParameter.php:347
/tmp/BetterReflection/test/unit/Reflection/ReflectionParameterTest.php:207

3) Roave\BetterReflectionTest\Reflection\ReflectionPropertyTest::testGetDocBlockTypeStrings with data set #0 ('privateProperty', array('int', 'float', '\stdClass'))
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindPropertyType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:232
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:216
/tmp/BetterReflection/test/unit/Reflection/ReflectionPropertyTest.php:130

4) Roave\BetterReflectionTest\Reflection\ReflectionPropertyTest::testGetDocBlockTypeStrings with data set #1 ('protectedProperty', array('bool', 'bool[]', 'bool[][]'))
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindPropertyType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:232
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:216
/tmp/BetterReflection/test/unit/Reflection/ReflectionPropertyTest.php:130

5) Roave\BetterReflectionTest\Reflection\ReflectionPropertyTest::testGetDocBlockTypeStrings with data set #2 ('publicProperty', array('string'))
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindPropertyType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:232
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:216
/tmp/BetterReflection/test/unit/Reflection/ReflectionPropertyTest.php:130

6) Roave\BetterReflectionTest\Reflection\ReflectionPropertyTest::testGetDocBlockTypes with data set #0 ('privateProperty', array('phpDocumentor\Reflection\Types\Integer', 'phpDocumentor\Reflection\Types\Float_', 'phpDocumentor\Reflection\Types\Object_'))
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindPropertyType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:232
/tmp/BetterReflection/test/unit/Reflection/ReflectionPropertyTest.php:153

7) Roave\BetterReflectionTest\Reflection\ReflectionPropertyTest::testGetDocBlockTypes with data set #1 ('protectedProperty', array('phpDocumentor\Reflection\Types\Boolean', 'phpDocumentor\Reflection\Types\Array_', 'phpDocumentor\Reflection\Types\Array_'))
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindPropertyType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:232
/tmp/BetterReflection/test/unit/Reflection/ReflectionPropertyTest.php:153

8) Roave\BetterReflectionTest\Reflection\ReflectionPropertyTest::testGetDocBlockTypes with data set #2 ('publicProperty', array('phpDocumentor\Reflection\Types\String_'))
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindPropertyType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:232
/tmp/BetterReflection/test/unit/Reflection/ReflectionPropertyTest.php:153

9) Roave\BetterReflectionTest\TypesFinder\FindParameterTypeTest::testNamespaceResolutionForProperty
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindParameterType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionParameter.php:347
/tmp/BetterReflection/src/Reflection/ReflectionParameter.php:330
/tmp/BetterReflection/test/unit/TypesFinder/FindParameterTypeTest.php:71

10) Roave\BetterReflectionTest\TypesFinder\FindPropertyTypeTest::testNamespaceResolutionForProperty
TypeError: array_filter() expects parameter 1 to be array, array given

/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:75
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:56
/tmp/BetterReflection/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php:28
/tmp/BetterReflection/src/TypesFinder/FindPropertyType.php:48
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:232
/tmp/BetterReflection/src/Reflection/ReflectionProperty.php:216
/tmp/BetterReflection/test/unit/TypesFinder/FindPropertyTypeTest.php:84

--

There was 1 failure:

1) Roave\BetterReflectionTest\SourceLocator\Type\PhpInternalSourceLocatorTest::testAllGeneratedStubsAreInSyncWithInternalReflectionClasses with data set "SoapServer" ('SoapServer')
SoapServer#setClass.args
Failed asserting that false is identical to true.

/tmp/BetterReflection/test/unit/SourceLocator/Type/PhpInternalSourceLocatorTest.php:431
/tmp/BetterReflection/test/unit/SourceLocator/Type/PhpInternalSourceLocatorTest.php:369
/tmp/BetterReflection/test/unit/SourceLocator/Type/PhpInternalSourceLocatorTest.php:301
/tmp/BetterReflection/test/unit/SourceLocator/Type/PhpInternalSourceLocatorTest.php:207

ERRORS!
Tests: 1965, Assertions: 27091, Errors: 10, Failures: 1, Skipped: 4, Incomplete: 15.
```
Tested with PHP 7.3 alpha4 and beta1.

The error itself looks fishy, there may be some weird bug in PHP itself.

**Comment by Majkl578** [comment]
This looks like some sort of internal zval corruption (not OPcache related).
 The following change makes array_filter-related failures pass:
```diff
diff --git a/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php b/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php
index 67e739d..d07c36b 100644
--- a/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php
+++ b/src/TypesFinder/PhpDocumentor/NamespaceNodeToReflectionTypeContext.php
@@ -66,7 +66,7 @@ class NamespaceNodeToReflectionTypeContext
     private function classAlikeUses(Namespace_ $namespace) : array
     {
         return array_filter(
-            $namespace->stmts ?? [],
+            ($namespace->stmts ?? []) + [],
             function (Node $node) : bool {
                 return (
                     $node instanceof Use_
```

**Comment by Ocramius** [comment]
This looks like a PHP 7.3 release blocker rather than an issue with this library.

**Comment by Majkl578** [comment]
Recompiled PHP In debug mode, crashes on internal assertion error:
```
$ /build/php/php-src/sapi/cli/php vendor/bin/phpunit test/unit/TypesFinder/FindPropertyTypeTest.php --filter testNamespaceRes
PHPUnit 7.3.1 by Sebastian Bergmann and contributors.

php: /build/php/php-src/Zend/zend_execute.c:233: _get_zval_ptr_tmp: Assertion `zval_get_type(&(*(ret))) != 10' failed.
Aborted
```
BT:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff783d2f1 in __GI_abort () at abort.c:79
#2  0x00007ffff7834a8a in __assert_fail_base (fmt=0x7ffff7988ec8 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x55555645b808 "zval_get_type(&(*(ret))) != 10", file=file@entry=0x55555645b7e0 "/build/php/php-src/Zend/zend_execute.c", line=line@entry=233, 
    function=function@entry=0x55555645de90 <__PRETTY_FUNCTION__.13265> "_get_zval_ptr_tmp") at assert.c:92
#3  0x00007ffff7834b02 in __GI___assert_fail (assertion=0x55555645b808 "zval_get_type(&(*(ret))) != 10", file=0x55555645b7e0 "/build/php/php-src/Zend/zend_execute.c", line=233, function=0x55555645de90 <__PRETTY_FUNCTION__.13265> "_get_zval_ptr_tmp") at assert.c:101
#4  0x0000555555c48a3d in _get_zval_ptr_tmp (var=96, should_free=0x7fffffffa3a0) at /build/php/php-src/Zend/zend_execute.c:233
#5  0x0000555555c774f2 in ZEND_COALESCE_SPEC_TMP_HANDLER () at /build/php/php-src/Zend/zend_vm_execute.h:18404
#6  0x0000555555cbb89c in execute_ex (ex=0x7ffff4c1f030) at /build/php/php-src/Zend/zend_vm_execute.h:57446
#7  0x0000555555cbf0e2 in zend_execute (op_array=0x7ffff4c85300, return_value=0x0) at /build/php/php-src/Zend/zend_vm_execute.h:60882
#8  0x0000555555bee6e7 in zend_execute_scripts (type=8, retval=0x0, file_count=3) at /build/php/php-src/Zend/zend.c:1562
#9  0x0000555555b5e33e in php_execute_script (primary_file=0x7fffffffdbb0) at /build/php/php-src/main/main.c:2630
#10 0x0000555555cc1b02 in do_cli (argc=5, argv=0x555556633240) at /build/php/php-src/sapi/cli/php_cli.c:997
#11 0x0000555555cc2a82 in main (argc=5, argv=0x555556633240) at /build/php/php-src/sapi/cli/php_cli.c:1390
```

Reported upstream: https://bugs.php.net/bug.php?id=76752

**Comment by asgrim** [comment]
Nice find @Majkl578 :grin:

**Comment by Majkl578** [comment]
The crash is fixed in 7.3.0beta3, but the one failure still remains, thus reopening. :)
```
1) Roave\BetterReflectionTest\SourceLocator\Type\PhpInternalSourceLocatorTest::testAllGeneratedStubsAreInSyncWithInternalReflectionClasses with data set "SoapServer" ('SoapServer')
SoapServer#setClass.args
Failed asserting that false is identical to true.

/tmp/BetterReflection/test/unit/SourceLocator/Type/PhpInternalSourceLocatorTest.php:431
/tmp/BetterReflection/test/unit/SourceLocator/Type/PhpInternalSourceLocatorTest.php:369
/tmp/BetterReflection/test/unit/SourceLocator/Type/PhpInternalSourceLocatorTest.php:301
/tmp/BetterReflection/test/unit/SourceLocator/Type/PhpInternalSourceLocatorTest.php:207
```

**Comment by asgrim** [comment]
Unsurprising. Stubs are the bane of this project :grin:
